### PR TITLE
Switch from arguments to an input structure for kubectl command

### PR DESCRIPTION
**Pull request by soltysh**
#### What type of PR is this?
/kind cleanup
/priority backlog
/sig cli

#### What this PR does / why we need it:
This PR:
1. introduces new `WithWrapConfigFn` function allowing to provider a wrapper for client config
2. switches kubectl from using multiple arguments to using a structure when constructing kubectl command.

This is alternative to https://github.com/kubernetes/kubernetes/pull/104926

#### Special notes for your reviewer:
/assign @eddiezane 
/cc @dovholuknf

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by dovholuknf** [comment]
I am happy to confirm that these changes, from the outside consumer point of view, are exactly what's needed by me. works perfectly and simplifies our code too as well as making the api cleaner.

**Comment by soltysh** [comment]
/hold
it looks like this PR breaks `-v6` I'll dig into it

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go
Comment (eddiezane) [inline staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go:395]: Do we want to allow chaining of these functions? This is really what I meant when I said this should be a slice.

```
func main() {
	flags := genericclioptions.NewConfigFlags(false)
	flags = flags.WithWrapConfigFn(func(config *rest.Config) *rest.Config {
		fmt.Println("wrap 1")
		return config
	})
	flags = flags.WithWrapConfigFn(func(config *rest.Config) *rest.Config {
		fmt.Println("wrap 2")
		return config
	})
	flags.ToRESTConfig()
}

// wrap 2
// wrap 1
// Unsure of the order we'd want here.
```

Or am I missing something here?

```suggestion
func (f *ConfigFlags) WithWrapConfigFn(wrapConfigFn func(*rest.Config) *rest.Config) *ConfigFlags {
	if f.WrapConfigFn == nil {
		f.WrapConfigFn = wrapConfigFn
	} else {
		old := f.WrapConfigFn
		f.WrapConfigFn = func(config *rest.Config) *rest.Config {
			return old(wrapConfigFn(config))
		}
	}
	return f
}
```

**Comment by dovholuknf** [comment]
fwiw, for me the change that broke the `-v6` flag was due to how the command is executed. prior it was simply created from the factory and then executed. now for `-v6` to function one needs to to invoke the command using `cli.Run`. the default `kubectl.go` file was updated already and works fine.

I adapted my code to use `cli.Run` and I can confirm `-v6` works for me.

**Comment by soltysh** [comment]
/hold cancel
the problem was elsewhere

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go
Comment (soltysh) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go:395]: Since `WrapConfigFn` both accepts and returns `*rest.Config` you can chain then together on your own, and pass that into `WithWrapConfigFn` that's why I don't think we need or want to make this a slice.

**Review by eddiezane — APPROVED** [review]

**Comment by dims** [comment]
the `cmd/` changes are straight forward and LGTM

/approve
/lgtm

**Comment by dims** [comment]
/assign @thockin @liggitt 

for `/cmd` approval (looks like we are light on approvers, gonna add myself)

**Comment by dims** [comment]
/approve
/lgtm

(for the record)

**Comment by soltysh** [comment]
> for `/cmd` approval (looks like we are light on approvers, gonna add myself)

I've went ahead and added cli-maintaners for the missing bits in https://github.com/kubernetes/kubernetes/pull/106451

**Pull request merged**
